doc/developer/design/20230418_persist_state_pubsub.md

```diff
@@ -0,0 +1,348 @@
+allows each end to push updates back and forth, with `environmentd` broadcasting out any changes it receives.
+
+## RPC Service
+
+```protobuf
+message ProtoPushDiff {
+  string shard_id = 1;
+  uint64 seqno = 2;
+  bytes diff = 3;
+}
+
+message ProtoSubscribe {
+  string shard = 1;
+}
+
+message ProtoUnsubscribe {
+  string shard = 1;
+}
+
+message ProtoPubSubMessage {
+  oneof message {
+    ProtoPushDiff push_diff = 1;
+    ProtoSubscribe subscribe = 2;
+    ProtoUnsubscribe unsubscribe = 3;
+  }
+}
+
+service ProtoPersistPubSub {
+  rpc PubSub (stream ProtoPubSubMessage) returns (stream ProtoPubSubMessage);
+}
+```
+
+The proposed topology would have each `clusterd` connect to `environmentd`'s RPC service. `clusterd` would subscribe to
+updates to any shards in its `PersistClientCache` (a singleton in both `environmentd` and `clusterd`). When any `clusterd`
+commits a change to a shard, it publishes the diff to `environmentd`, which tracks which connections are subscribed to which
+shards, and broadcasts out the diff to the interested subscribers. From there, each subscriber would [apply the diff](#applying-received-updates).
+Note that a `clusterd` would only subscribe to a shard if it is actively reading it, e.g. for a materialized view, `SELECT`,
+or a `SUBSCRIBE`, so the number of broadcasted messages would be proportional to the number of ongoing read operations,
+and not the total number of shards.
+
+`environmentd` is chosen to host the RPC service as it necessarily holds a handle to each shard (and therefore is always
+interested in all diffs), and to align with the existing topology of the Controller, rather than introduce the novel
+complexity of point-to-point connections between unrelated `clusterd` processes or a net-new dedicated process. With this
+implementation we would anticipate the end-to-end latency between a writer in one `clusterd` pushing an update to a reader
+in another `clusterd` receiving that update to be in the 1-10ms range.
+
+### Service Discovery
+
+`clusterd` will need to know the address of the `environmentd` process in order to connect to the PubSub server. Locally,
+this will be done through a standard `127.0.0.1:6879` port and passed in to `clusterd` via command line argument.
+
+In Kubernetes, we will create a new headless `Service` in the `environment-controller` that routes to the pubsub port
+(likely 6879) of the current generation of `envd`:
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  labels:
+    materialize.cloud/environment-name: env-name
+    materialize.cloud/organization-id: org-name
+  name: persist-pubsub-<envd-generation>
+  namespace: env-name
+spec:
+  selector:
+    materialize.cloud/app: environmentd
+    # ensure this service only routes to a specific generation of `envd`
+    materialize.cloud/envd-generation: <envd-generation>
+  type: ClusterIP
+  clusterIP: None
+  ports:
+  - name: pubsub
+    port: 6879
+    protocol: TCP
+    targetPort: 6879
+```
```

Comment:
You'll need to teach cloudtest to do this too! Which, is probably fine, but points further in the direction of "we haven't gotten the abstraction boundary right between cloud infra, the database, and our testing infra."